### ISSUE TITLE
False CA2213

### ISSUE THREAD

**Issue by jherby2k** [dotnet/roslyn-analyzers#2413]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.2

#### Diagnostic ID

[CA2213](https://docs.microsoft.com/en-ca/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed)

#### Repro steps

I'm getting false positives if an IDisposable field is not disposed, even if the object was not locally created. It only happens if the class implements IDisposable.

```c#
    sealed class StreamTest : IDisposable
    {
        readonly Stream _stream;

        public StreamTest(Stream stream)
        {
            _stream = stream;
        }

        public void Dispose()
        {
            // _stream shouldn't get disposed in this class!
        }
    }
```

#### Expected behavior

No warning

#### Actual behavior

warning

**Comment by mavasani** [comment]
@jherby2k This is actually by design, and the diagnostic fires for this case in both binary FxCop and the ported rule in FxCop analyzers. The core dispose analysis for both the implementations special cases constructor parameters of below 4 types to have a dispose ownership transfer:

1. System.IO.Stream
2. System.IO.TextReader
3. System.IO.TextWriter
4. System.Resources.IResourceReader

For majority of code bases, passes an object of this type to a constructor and then assigning it to a field indicates a dispose ownership transfer to the newly constructed type. So, [CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope?view=vs-2019) will not be fired for streams that are created and passed to an object creation as the creating method no longer owns the dispose ownership for that stream. However, CA2213 will fire due to dispose ownership transfer, but missing dispose.

**Comment by mavasani** [comment]
I also filed doc bug https://github.com/MicrosoftDocs/visualstudio-docs/issues/3198 to document this.

**Comment by jherby2k** [comment]
Okay, but why does CA2213 only fire if the constructed type already implements IDisposable? Remove that from my example and the warning disappears.

**Comment by mavasani** [comment]
@jherby2k That is the primary intent of this rule: https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed?view=vs-2019#cause

**Comment by jherby2k** [comment]
No, that would be the primary intent of a rule titled "Disposable fields should be disposed, unless you don't feel like implementing IDisposable."

Does not trigger CA2213:

```c#
    sealed class StreamTest
    {
        readonly Stream _stream;

        public StreamTest(Stream stream)
        {
            _stream = stream;
        }
    }
```

**Comment by mavasani** [comment]
What method(s) should be analyzed to expect dispose invocations on the field?

**Comment by mavasani** [comment]
Note CA2213 descriptions states: `A type that implements System.IDisposable declares fields that are of types that also implement IDisposable. The Dispose method of the field is not called by the Dispose method of the declaring type.`

**Comment by jherby2k** [comment]
Interesting. So this rule by checks for some things that aren't useful and doesn't check for some things that would be useful. By design. OK!
